### PR TITLE
Multi project build structure

### PR DESCRIPTION
**Pull request by kensipe**
This is for the multi-project build.  It is the minimum amount of work that is functional and logical.   Once this has been merged there are few things that need to change IMO.   The candidate changes:

1. in the common project, there is a SchedulerConfig... it should be broken up or renamed.  probably both.
2. there are classes in the same packages in different projects.  this isn't a problem, it just feels off.
3. we need to remove the build-hdfs script and move that into gradle build.

### CODE REVIEW

**Comment by elingg** [comment]
Thanks @kensipe! Will review this when I have a chance. @adam-mesos is an optional reviewer.

**Comment by kensipe** [comment]
sweet!

**Comment by elingg** [comment]
Frist comment- the build directory structure is a bit confusing. If we could put the "cached" artifacts in a sub dir of build as we discussed @kensipe , it will be cleaner. In the build directory, I would like to be able to easily find my executor and scheduler tarballs and not see duplicate copies.

**Comment by kensipe** [comment]
@elingg cleaned up some docs and changed the build dir.  It has a cache dir now.  you'll want to do a clean before running the build again.  I think we are good if you are good.

**Comment by kensipe** [comment]
and now tested on GCE.  I think we are good.

File: README.md [hunk 1]
```diff
@@ -17,6 +17,9 @@ Building HDFS-Mesos
 2. Run `./bin/build-hdfs nocompile` to skip the `gradlew clean package` step and just re-bundle the binaries.
 3. To remove the project build output and downloaded binaries, run `./bin/build-hdfs clean`
 
+**NOTE:** The build process builds the artifacts under the `$PROJ_DIR/build` directory.  A number of zip and tar files will be there which are cached for faster subsequent builds.   The tarball used for installation is hdfs-mesos-x.x.x.tgz which contains the scheduler and the executor to be distributed.
+
+
 Installing HDFS-Mesos on your Cluster
 --------------------------
```
Comment (elingg) [inline README.md]: We also want to note here to upload hdfs-mesos-*.tgz, which is under the build directory. Users may not realize that this file is now under the build directory.

**Comment by elingg** [comment]
Thanks, @kensipe, I prefer this structure. It appears to be much cleaner. Taking a look at the rest of the changes now.

File: bin/build-hdfs [hunk 2]
```diff
@@ -10,86 +16,117 @@ IMPALA_VER=cdh5.3.1-release
...
 
 # Remove cached binaries and exit
 if [ "$1" == "clean" ]; then
-    rm -f $HADOOP_ZIP
-    rm -rf $HADOOP_DIR
-    rm -rf $NATIVE
-    rm -rf $DIST
-    rm -f $DIST.tgz
+    rm -Rf $BUILD_DIR
     $PROJ_DIR/gradlew clean
```
Comment (elingg) [inline bin/build-hdfs]: Why are we using -Rf instead of -rf here? It should have the same functionality, just curious.

File: bin/build-hdfs [hunk 3]
```diff
@@ -10,86 +16,117 @@ IMPALA_VER=cdh5.3.1-release
...
     exit 0
 fi
 
 # Build and package hdfs-mesos project
```
Comment (elingg) [inline bin/build-hdfs]: If the option is NOT "nocompile," we should run ```gradlew clean shadowJar``` to make sure we are doing a clean compile. Why was this changed?

File: bin/build-hdfs [hunk 4]
```diff
@@ -10,86 +16,117 @@ IMPALA_VER=cdh5.3.1-release
...
     echo "Downloading $HADOOP_URL"
-    wget $HADOOP_URL || exit
+    wget -P $BUILD_CACHE_DIR $HADOOP_URL || exit
 else
     echo "($HADOOP_ZIP already exists, skipping dl)"
 fi
 
 # Extract hadoop
-if [ ! -d $HADOOP_DIR ]; then
-    echo "Extracting $HADOOP_ZIP"
```
Comment (elingg) [inline bin/build-hdfs]: a few spacing issues here

File: bin/build-hdfs [hunk 5]
```diff
@@ -10,86 +16,117 @@ IMPALA_VER=cdh5.3.1-release
...
     tar xf $HADOOP_ZIP
+	cd -
 else
     echo "($HADOOP_DIR already exists, skipping extract)"
 fi
 
 # Get native libraries
-if [ ! -d $NATIVE ]; then
+if [ ! -d $BUILD_CACHE_DIR/$NATIVE ]; then
     echo "Downloading and unpacking native libs"
```
Comment (elingg) [inline bin/build-hdfs]: Similar question here on the spacing. Why are we indenting cd?

File: build.gradle [hunk 6]
```diff
@@ -1,83 +1,79 @@
...
 
-repositories {
-    mavenLocal()
-    mavenCentral()
-}
+    group = "com.apache.mesos.hdfs"
+    version = "0.1.1"
 
-dependencies {
+    ext {
```
Comment (elingg) [inline build.gradle]: We should be able to upgrade to 0.22.0 now.

**Comment by elingg** [comment]
This is looking good, but I left some comments to be addressed. Any final objections or comments, @adam-mesos?

File: build.gradle (hunk 6 above)
Comment (adam-mesos) [inline build.gradle]: 0.22.1, please

Comment (elingg) [inline build.gradle]: opps yes, 0.22.1!

File: README.md (hunk 1 above)
Comment (kensipe) [inline README.md]: the details for uploading for install are in the next section of the readme.  I've updated that to be clear and to make it consistent with the new build process.

File: bin/build-hdfs (hunk 2 above)
Comment (kensipe) [inline bin/build-hdfs]: I switched it to be consistent with the past... its just an old habit

File: bin/build-hdfs (hunk 3 above)
Comment (kensipe) [inline bin/build-hdfs]: it can be necessary to do this with maven :), but gradle uses a hash to check input changes and does incremental builds.    doing a clean defeats that and shouldn't add any value.   Is there a concern?

File: bin/build-hdfs (hunk 4 above)
Comment (kensipe) [inline bin/build-hdfs]: I think this is an older version... this isn't the current code for that file.

**Comment by kensipe** [comment]
@elingg Its ready to go... all feedback resolved.

File: bin/build-hdfs (hunk 3 above)
Comment (elingg) [inline bin/build-hdfs]: I do have concerns with incremental builds and bugs in the build tools related to this. If you have full confidence in gradle (I don't have full confidence in maven due to past dependency issues), this is ok....

File: README.md [hunk 7]
```diff
@@ -17,9 +17,12 @@ Building HDFS-Mesos
...
 3. To remove the project build output and downloaded binaries, run `./bin/build-hdfs clean`
 
+**NOTE:** The build process builds the artifacts under the `$PROJ_DIR/build` directory.  A number of zip and tar files are cached under the `cache` directory for faster subsequent builds.   The tarball used for installation is hdfs-mesos-x.x.x.tgz which contains the scheduler and the executor to be distributed.
+
+
 Installing HDFS-Mesos on your Cluster
 --------------------------
-1. Upload `hdfs-mesos-*.tgz` to a node in your Mesos cluster.
+1. Upload `hdfs-mesos-*.tgz` to a node in your Mesos cluster.   (which is built to `$PROJ_DIR/build/hdfs-mesos-x.x.x.tgz`)
 2. Extract it with `tar zxvf hdfs-mesos-*.tgz`
```
Comment (elingg) [inline README.md]: nit: The period should be at the end and we should say $PROJ_DIR/build/hdfs-mesos-*.tgz instead of $PROJ_DIR/build/hdfs-mesos-x.x.x.tgz for consistency.

**Comment by elingg** [comment]
Thanks @kensipe. One nit in the docs. Once that is addressed, you can feel free to merge since you have tested on GCE and with DCOS.
